query/control/memory.go

```diff
@@ -0,0 +1,132 @@
+package control
+
+import (
+	"errors"
+	"sync/atomic"
+
+	"github.com/influxdata/flux/memory"
+)
+
+type memoryManager struct {
+	// initialBytesQuotaPerQuery is the initial amount of memory
+	// allocated for each query. It does not count against the
+	// memory pool.
+	initialBytesQuotaPerQuery int64
+
+	// memoryBytesQuotaPerQuery is the maximum amount of memory
+	// that may be allocated to each query.
+	memoryBytesQuotaPerQuery int64
+
+	// unusedMemoryBytes is the amount of memory that may be used
+	// when a query requests more memory.
+	unusedMemoryBytes int64
+}
+
+// createAllocator will construct an allocator and memory manager
+// for the given query.
+func (c *Controller) createAllocator(q *Query) {
+	q.memoryManager = &queryMemoryManager{
+		m:     c.memory,
+		limit: c.memory.initialBytesQuotaPerQuery,
+	}
+	q.alloc = &memory.Allocator{
+		// Use an anonymous function to ensure the value is copied.
+		Limit:   func(v int64) *int64 { return &v }(q.memoryManager.limit),
+		Manager: q.memoryManager,
+	}
+}
+
+// queryMemoryManager is a memory manager for a specific query.
+type queryMemoryManager struct {
+	m     *memoryManager
+	limit int64
+	given int64
+}
+
+// RequestMemory will determine if the query can be given more memory
+// when it is requested.
+//
+// Note: This function accesses the memoryManager whose attributes
+// may be modified concurrently. Atomic operations are used to keep
+// it lockless. The data associated with this specific query are only
+// invoked from within a lock so they are safe to modify.
+// Second Note: The errors here are discarded anyway so don't worry
+// too much about the specific message or structure.
+func (q *queryMemoryManager) RequestMemory(want int64) (got int64, err error) {
```

Comment:
Does this method have a test case? I don't see where RequestMemory is used. Is that planned as a separate PR?

Reply:
I see its used on the Flux side of things https://github.com/influxdata/flux/pull/1922/files#diff-576d3093e5f774162ca06b1345b6ed07R191